### PR TITLE
feat(object): add zipObjWith

### LINKED ISSUE #1129: zipObjWith
**Is your feature request related to a problem? Please describe.**

Same as https://ramdajs.com/docs/#zipObj but accepts a binary function as the first argument. This binary function returns an object which gets merged into final resulting object.

**Describe the solution you'd like**

```js
RA.zipObjWith((value, key) => [key, `${key}${value + 1}`]), ['a', 'b', 'c'], [1, 2, 3]); // => {a: 'a2', b: 'b3', c: 'c4'}
```

**Describe alternatives you've considered**

```js
RA.zipObjWith((value, key) => ({[key]: value}), ['a', 'b', 'c'], [1, 2, 3]); // => {'a': 1, 'b': 2, 'c': 3}
RA.zipObjWith((value, key) => `key${value + 1}`, ['a', 'b', 'c'], [1, 2, 3]); // => {'a': a2, 'b': b3, 'c': c4}
```

Before we go for implementation we have to decide which implementation makes more sense and have higher value.

### PR DESCRIPTION
**Pull request by thomasmichaelwallace**
Closes #1129

### CODE REVIEW

File: src/zipObjWith.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * Creates a new object out of a list of keys and a list of values by applying the function
+ * to each equally-positioned pair in the lists.
+ * Key/value pairing is truncated to the length of the shorter of the two lists.
+ *
+ * @func zipObjWith
+ * @memberOf RA
+ * @category Object
+ * @since {@link https://char0n.github.io/ramda-adjunct/2.22.0|v2.22.0}
+ * @sig Functor f => (v, k) => [c, d] -> [k] -> [v] -> { c: d }
```
Comment (char0n) [inline src/zipObjWith.js]: Align this signature with `unzipObjWith`. This function is not a functor as I pointed out in that PR.

File: src/zipObjWith.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * @param {Function} fn The function to transform each value-key pair
+ * @param {Array} keys Array to transform into the properties on the output object
+ * @param {Array} values Array to transform into the values on the output object
+ * @return {Object}  The object made by pairing up and transforming same-indexed elements of `keys` and `values`.
+ * @see {@link https://ramdajs.com/docs/#zipObj}
```
Comment (char0n) [inline src/zipObjWith.js]: `@see {@link https://ramdajs.com/docs/#zipObj|zipObj}`

File: test/zipObjWith.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import { assert } from 'chai';
+import * as R from 'ramda';
+
+import * as RA from '../src';
+
+describe('unzipObjWith', function() {
```
Comment (char0n) [inline test/zipObjWith.js]: zipObjWith

File: types/test/zipObjWith.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import * as RA from 'ramda-adjunct';
+
+RA.zipObjWith((v, t) => [t, 1], ['a'], [2]); // $ExpectType {}
```
Comment (char0n) [inline types/test/zipObjWith.ts]: More tests here, see `unzipObjWith` PR

File: types/index.d.ts [hunk 5]
```diff
@@ -820,6 +820,12 @@ declare namespace RamdaAdjunct {
         flattenProp(prop: string | number, obj: object): object;
         flattenProp(prop: string | number): (obj: object) => object;
 
+        /**
+         * Creates a new object out of a list of keys and a list of values by applying the function
+         * to each equally-positioned pair in the lists.
+         */
+        zipObjWith<T, U, V>(fn: (value: T, key: U) => [string, V], keys: U[], values: T[]): {};
```
Comment (char0n) [inline types/index.d.ts]: Align with unzipObjWith signature + add currying support

File: types/index.d.ts [hunk 6]
```diff
@@ -820,6 +820,12 @@ declare namespace RamdaAdjunct {
         flattenProp(prop: string | number, obj: object): object;
         flattenProp(prop: string | number): (obj: object) => object;
 
+        /**
+         * Creates a new object out of a list of keys and a list of values by applying the function
```
Comment (char0n) [inline types/index.d.ts]: Description here doesn't equal the description in function definition.

File: test/zipObjWith.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  it('unzips an object into key/value arrays applying a value/key transform', function() {
```
Comment (char0n) [inline test/zipObjWith.js]: Use tests conventions already utilized in `unzipObjWith`. `should/context/specify` etc...

File: src/zipObjWith.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * Creates a new object out of a list of keys and a list of values by applying the function
+ * to each equally-positioned pair in the lists.
+ * Key/value pairing is truncated to the length of the shorter of the two lists.
+ *
+ * @func zipObjWith
+ * @memberOf RA
+ * @category Object
+ * @since {@link https://char0n.github.io/ramda-adjunct/2.22.0|v2.22.0}
+ * @sig (v, k) -> [k, v] -> [k] -> [v] -> { c: d }
```
Comment (char0n) [inline src/zipObjWith.js]: Signature dosn't seem right. It's not apparent where `c` and `d` comes from. So I'd opt for one of the following:
```js
@sig (v, k) -> [k, v] -> [k] -> [v] -> { k: v }
```
or
```js
@sig (b, a) -> [k, v] -> [a] -> [b] -> { k: v }
```

I think the second signature captures reality more. We have two list of arbitrary values, we apply the transformation function on those values which gives us `(k, v)` pairs, which are then transformed to `{ k: v }` object. It's implied by javascript implementation that `k is String`, but `a is *`. `k` is created by transforming `a` to String.

Comment (thomasmichaelwallace) [reply src/zipObjWith.js]: Agreed- I think `c` and `d` are there from a previous attempt!

**Comment by thomasmichaelwallace** [comment]
I think conflicts will resolve with second build... - first build was intermediate push.

**Comment by char0n** [comment]
Waiting for tests. I'll sqash + merge the PR to avoid conflicts. No action required from your side. Again, thank you for contributing!

**Pull request merged**